src/display/node_utils.js

```diff
@@ -27,56 +27,90 @@ if (typeof PDFJSDev !== "undefined" && PDFJSDev.test("MOZCENTRAL")) {
+  // eslint-disable-next-line no-var
+  var packageMap = null;
 
-if (typeof PDFJSDev !== "undefined" && !PDFJSDev.test("SKIP_BABEL")) {
-  (function checkDOMMatrix() {
-    if (globalThis.DOMMatrix || !isNodeJS) {
-      return;
-    }
-    const DOMMatrix = canvas?.DOMMatrix;
+  const loadPackages = async () => {
+    // Native packages.
+    const fs = await __non_webpack_import__("fs"),
+      http = await __non_webpack_import__("http"),
+      https = await __non_webpack_import__("https"),
+      url = await __non_webpack_import__("url");
 
-    if (DOMMatrix) {
-      globalThis.DOMMatrix = DOMMatrix;
-    } else {
-      warn("Cannot polyfill `DOMMatrix`, rendering may be broken.");
+    // Optional, third-party, packages.
+    let canvas, path2d;
+    if (typeof PDFJSDev !== "undefined" && !PDFJSDev.test("SKIP_BABEL")) {
+      try {
+        canvas = await __non_webpack_import__("canvas");
+      } catch {}
+      try {
+        path2d = await __non_webpack_import__("path2d");
+      } catch {}
     }
-  })();
 
-  (function checkPath2D() {
-    if (globalThis.Path2D || !isNodeJS) {
-      return;
-    }
-    const CanvasRenderingContext2D = canvas?.CanvasRenderingContext2D;
-    const applyPath2DToCanvasRenderingContext =
-      path2d?.applyPath2DToCanvasRenderingContext;
-    const Path2D = path2d?.Path2D;
-
-    if (
-      CanvasRenderingContext2D &&
-      applyPath2DToCanvasRenderingContext &&
-      Path2D
-    ) {
-      applyPath2DToCanvasRenderingContext(CanvasRenderingContext2D);
-      globalThis.Path2D = Path2D;
-    } else {
-      warn("Cannot polyfill `Path2D`, rendering may be broken.");
+    return new Map(Object.entries({ fs, http, https, url, canvas, path2d }));
+  };
+
+  loadPackages().then(
+    map => {
+      packageMap = map;
+      packageCapability.resolve();
+
+      if (typeof PDFJSDev === "undefined" || PDFJSDev.test("SKIP_BABEL")) {
+        return;
+      }
+      if (!globalThis.DOMMatrix) {
+        const DOMMatrix = map.get("canvas")?.DOMMatrix;
+
+        if (DOMMatrix) {
+          globalThis.DOMMatrix = DOMMatrix;
+        } else {
+          warn("Cannot polyfill `DOMMatrix`, rendering may be broken.");
+        }
+      }
+      if (!globalThis.Path2D) {
+        const CanvasRenderingContext2D =
+          map.get("canvas")?.CanvasRenderingContext2D;
+        const applyPath2DToCanvasRenderingContext =
+          map.get("path2d")?.applyPath2DToCanvasRenderingContext;
+        const Path2D = map.get("path2d")?.Path2D;
+
+        if (
+          CanvasRenderingContext2D &&
+          applyPath2DToCanvasRenderingContext &&
+          Path2D
+        ) {
+          applyPath2DToCanvasRenderingContext(CanvasRenderingContext2D);
+          globalThis.Path2D = Path2D;
+        } else {
+          warn("Cannot polyfill `Path2D`, rendering may be broken.");
+        }
+      }
+    },
+    reason => {
+      warn(`loadPackages: ${reason}`);
+
+      packageMap = new Map();
```

Comment:
Given that I cannot really see a scenario in which loading native modules can fail, and the optional modules are already handled separately, is there actually a way to hit this `catch` branch? Even if we do, some call sites of `get(name)` for the seem to assume that a valid package is returned, but if loading fails the map will be empty and we'll likely see errors occurring later on in the places where `get(name)` is called.

Shouldn't we just raise a fatal exception here instead of warning, since if we get here I don't really see how we can continue gracefully here without breaking later on in the chain?

Reply:
We probably shouldn't ever hit this branch, for the reasons you mention, but it still seemed "best" to handle any errors here (however unlikely they might be).

Depending on the PDF document you open *and* how, there might still be cases where none of the Node.js packages/polyfills are actually necessary.
For example, let's assume that you load a "simple" PDF document from a TypedArray in a Node.js environment and want to extract the textContent. Assuming that the fonts are *all embedded*, and don't require CMaps, then none of the Node.js dependencies should be necessary in this case and failing to load should thus not be an issue.

Perhaps I'm overthinking this, but the above was basically my line of thinking when I wrote that code.

Reply:
Good point; I thought we were always bound to hit at least one of those call sites. If not, then I'm happy with the current code indeed.